schemacrawler crashes on Oracle - java.sql.SQLSyntaxErrorException: ORA-00920: invalid relational operator

# Bug Report

**Issue**

Please explain the issue briefly
* I need to investigate an ancient Oracle database, the bundled JDBC driver leads to ArrayIndexOutOfBoundsException
* I downloaded the ojdbc6.jar from https://download.oracle.com/otn/utilities_drivers/jdbc/11204/ojdbc6.jar and renamed the bundled ojdbc8_g-19.7.0.0.jar to ojdbc8_g-19.7.0.0.jar_
* I start schemacrawler with 
```
~/java/schemacrawler/schemacrawler.sh --server=oracle --host=10.23.15.212 --port=1521 --urlx=oracle.jdbc.timezoneAsRegion=false --database=WASS --schemas=WASS --routines= --sequences=".*" --c=schema --user=wass --password=passwd --info-level=maximum --log-level=ALL --command=schema &> debug.log
```
* Result is Error: Exception retrieving table information: ORA-00920: invalid relational operator

Debug log: [debug.log.gz](https://github.com/schemacrawler/SchemaCrawler/files/4907437/debug.log.gz)

**Environment**

* SchemaCrawler 16.9.2
* Linux 5.4.0-40-generic
* Ubuntu OpenJDK 64-Bit Server VM 11.0.7+10-post-Ubuntu-3ubuntu1
* Oracle9i Enterprise Edition 9.2.0.1.0
* JDBC driver ojdbc6.jar

Thanks for the swift reply!
When I remove the plugin from the classpath, schemacrawler itself is unhappy:

```
Jul 11, 2020 9:43:07 PM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logFullStackTrace
SEVERE: Cannot connect to database
java.lang.RuntimeException: Cannot connect to database
	at schemacrawler.tools.commandline.command.ConnectCommand.run(ConnectCommand.java:127)
	at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.executeCommandLine(SchemaCrawlerCommandLine.java:141)
	at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.execute(SchemaCrawlerCommandLine.java:72)
	at schemacrawler.Main.main(Main.java:85)
Caused by: schemacrawler.schemacrawler.SchemaCrawlerException: SchemaCrawler database plugin should be on the CLASSPATH for <jdbc:oracle:thin:wass/passwd@//10.23.15.212:1521/WASS;oracle.jdbc.timezoneAsRegion=false>
	at schemacrawler.tools.databaseconnector.UnknownDatabaseConnector.newDatabaseConnectionSource(UnknownDatabaseConnector.java:85)
	at schemacrawler.tools.commandline.command.ConnectCommand.createDataSource(ConnectCommand.java:180)
	at schemacrawler.tools.commandline.command.ConnectCommand.run(ConnectCommand.java:119)
	... 3 more
```

While I did not check the source, I assume that the `oracle` type is recognized and a plugin is assumed to be available. (When using a wrong type like `foo` schemacrawler shows a different error message.)

I checked: REGEXP_LIKE is not support in Oracle 9, only in Oracle 10. There is a polyfill available, though:

```
CREATE OR REPLACE FUNCTION regexp_like (source_char     IN VARCHAR2
                                      , pattern         IN VARCHAR2
                                      , match_parameter IN VARCHAR2 DEFAULT NULL)
RETURN INTEGER
AS
/*************************************************************************
 * Program  : regexp_like
 * Version  : 1.0
 * Author   : Philip Moore
 * Date     : 20-JUN-2009 Anno Domini
 * Purpopse : This provides a pseudo "REGEXP_LIKE" operator for Oracle 9iR2
 * Warnings : Do NOT compile in an Oracle 10GR2 database (or above)!
 *************************************************************************/
  -- Variables
  l_return INTEGER;
BEGIN
   IF owa_pattern.match (line  => source_char
                       , pat   => pattern
                       , flags => match_parameter) THEN
      l_return := 1;
   ELSE
      l_return := 0;
   END IF;

   RETURN l_return;
END regexp_like;
/
```
I added it using the same schema I am targeting with schemacrawler, but it seems that the function was not available, since the original error persisted.

--- a/schemacrawler-commandline/src/main/java/schemacrawler/tools/commandline/command/ConnectCommand.java
+++ b/schemacrawler-commandline/src/main/java/schemacrawler/tools/commandline/command/ConnectCommand.java
@@ -122,7 +122,11 @@ public class ConnectCommand
       loadSchemaRetrievalOptionsBuilder(databaseConnector);
 
     }
-    catch (final SchemaCrawlerException | SQLException e)
+    catch (final SchemaCrawlerException e)
+    {
+      throw new RuntimeException(e.getMessage(), e);
+    }
+    catch (final SQLException e)
     {
       throw new RuntimeException("Cannot connect to database", e);
     }

--- a/schemacrawler-distrib/src/examples/resources/_testdb/createtestschema.sh
+++ b/schemacrawler-distrib/src/examples/resources/_testdb/createtestschema.sh
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash
 SC_DIR=$(dirname "$0")
-java -cp "$(echo "$SC_DIR"/lib/*.jar | tr ' ' ':'):." schemacrawler.testdb.TestSchemaCreatorMain "$*"
+java -cp "$(echo "$SC_DIR"/lib/*.jar | tr ' ' ':'):." schemacrawler.testdb.TestSchemaCreatorMain "$@"

--- a/schemacrawler-hsqldb/src/test/java/schemacrawler/integration/test/MatchSchemaRetrievalOptionsTest.java
+++ b/schemacrawler-hsqldb/src/test/java/schemacrawler/integration/test/MatchSchemaRetrievalOptionsTest.java
@@ -1,0 +1,112 @@
+/*
+========================================================================
+SchemaCrawler
+http://www.schemacrawler.com
+Copyright (c) 2000-2020, Sualeh Fatehi <sualeh@hotmail.com>.
+All rights reserved.
+------------------------------------------------------------------------
+
+SchemaCrawler is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+
+SchemaCrawler and the accompanying materials are made available under
+the terms of the Eclipse Public License v1.0, GNU General Public License
+v3 or GNU Lesser General Public License v3.
+
+You may elect to redistribute this code under any of these licenses.
+
+The Eclipse Public License is available at:
+http://www.eclipse.org/legal/epl-v10.html
+
+The GNU General Public License v3 and the GNU Lesser General Public
+License v3 are available at:
+http://www.gnu.org/licenses/
+
+========================================================================
+*/
+package schemacrawler.integration.test;
+
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static schemacrawler.utility.SchemaCrawlerUtility.matchSchemaRetrievalOptions;
+
+import java.sql.Connection;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import schemacrawler.schemacrawler.DatabaseServerType;
+import schemacrawler.schemacrawler.SchemaRetrievalOptions;
+import schemacrawler.test.utility.TestDatabaseConnectionParameterResolver;
+
+@ExtendWith(TestDatabaseConnectionParameterResolver.class)
+public class MatchSchemaRetrievalOptionsTest
+{
+
+  @DisplayName(
+    "\"SC_WITHOUT_DATABASE_PLUGIN\" is set to \"hsqldb\" - use \"unknown\" plugin")
+  @Test
+  public void matchSchemaRetrievalOptions1(final Connection connection)
+    throws Exception
+  {
+    try
+    {
+      System.setProperty("SC_WITHOUT_DATABASE_PLUGIN", "hsqldb");
+      final SchemaRetrievalOptions schemaRetrievalOptions =
+        matchSchemaRetrievalOptions(connection);
+      final DatabaseServerType databaseServerType =
+        schemaRetrievalOptions.getDatabaseServerType();
+      assertThat(databaseServerType.isUnknownDatabaseSystem(), is(true));
+    }
+    finally
+    {
+      System.clearProperty("SC_WITHOUT_DATABASE_PLUGIN");
+    }
+  }
+
+  @DisplayName(
+    "\"SC_WITHOUT_DATABASE_PLUGIN\" is set to \"newdb\" - use \"hsqldb\" plugin")
+  @Test
+  public void matchSchemaRetrievalOptions2(final Connection connection)
+    throws Exception
+  {
+    try
+    {
+      System.setProperty("SC_WITHOUT_DATABASE_PLUGIN", "newdb");
+      final SchemaRetrievalOptions schemaRetrievalOptions =
+        matchSchemaRetrievalOptions(connection);
+      final DatabaseServerType databaseServerType =
+        schemaRetrievalOptions.getDatabaseServerType();
+      assertThat(databaseServerType.getDatabaseSystemIdentifier(),
+                 is("hsqldb"));
+    }
+    finally
+    {
+      System.clearProperty("SC_WITHOUT_DATABASE_PLUGIN");
+    }
+  }
+
+  @DisplayName("\"SC_WITHOUT_DATABASE_PLUGIN\" is not set - use default plugin")
+  @Test
+  public void matchSchemaRetrievalOptions0(final Connection connection)
+    throws Exception
+  {
+    try
+    {
+      System.setProperty("SC_WITHOUT_DATABASE_PLUGIN", "newdb");
+      final SchemaRetrievalOptions schemaRetrievalOptions =
+        matchSchemaRetrievalOptions(connection);
+      final DatabaseServerType databaseServerType =
+        schemaRetrievalOptions.getDatabaseServerType();
+      assertThat(databaseServerType.getDatabaseSystemIdentifier(),
+                 is("hsqldb"));
+    }
+    finally
+    {
+      System.clearProperty("SC_WITHOUT_DATABASE_PLUGIN");
+    }
+  }
+
+}

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
@@ -58,8 +58,7 @@ final class UnknownDatabaseConnector
   {
     super(DatabaseServerType.UNKNOWN,
           new EmptyInputResource(),
-          (informationSchemaViewsBuilder, connection) -> informationSchemaViewsBuilder.fromResourceFolder(
-            "/db2.information_schema"));
+          (informationSchemaViewsBuilder, connection) -> {} );
   }
 
   @Override

--- a/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/tools/databaseconnector/UnknownDatabaseConnector.java
@@ -69,7 +69,7 @@ final class UnknownDatabaseConnector
       super.newDatabaseConnectionSource(databaseConnectorOptions);
 
     final String withoutDatabasePlugin = getSystemConfigurationProperty(
-      "SC_WITHOUT_DATABASE_PLUGIN",
+      "SC_IGNORE_MISSING_DATABASE_PLUGIN",
       Boolean.FALSE.toString());
     if (!Boolean.valueOf(withoutDatabasePlugin))
     {

--- a/schemacrawler-tools/src/main/java/schemacrawler/utility/SchemaCrawlerUtility.java
+++ b/schemacrawler-tools/src/main/java/schemacrawler/utility/SchemaCrawlerUtility.java
@@ -29,6 +29,9 @@ http://www.gnu.org/licenses/
 package schemacrawler.utility;
 
 
+import static schemacrawler.schemacrawler.Config.getSystemConfigurationProperty;
+import static sf.util.Utility.isBlank;
+
 import java.sql.Connection;
 import java.sql.ResultSet;
 import java.util.logging.Level;
@@ -37,6 +40,7 @@ import schemacrawler.crawl.ResultsCrawler;
 import schemacrawler.crawl.SchemaCrawler;
 import schemacrawler.schema.Catalog;
 import schemacrawler.schema.ResultsColumns;
+import schemacrawler.schemacrawler.DatabaseServerType;
 import schemacrawler.schemacrawler.SchemaCrawlerException;
 import schemacrawler.schemacrawler.SchemaCrawlerOptions;
 import schemacrawler.schemacrawler.SchemaCrawlerSQLException;
@@ -77,12 +81,27 @@ public final class SchemaCrawlerUtility
       DatabaseConnectorRegistry.getDatabaseConnectorRegistry();
     final DatabaseConnector dbConnector =
       registry.lookupDatabaseConnector(connection);
-    LOGGER.log(Level.INFO,
-               "Using database plugin for "
-               + dbConnector.getDatabaseServerType());
+    final DatabaseServerType databaseServerType =
+      dbConnector.getDatabaseServerType();
+    LOGGER.log(Level.INFO, "Using database plugin for " + databaseServerType);
 
-    final SchemaRetrievalOptionsBuilder schemaRetrievalOptionsBuilder =
-      dbConnector.getSchemaRetrievalOptionsBuilder(connection);
+    final String withoutDatabasePlugin =
+      getSystemConfigurationProperty("SC_WITHOUT_DATABASE_PLUGIN", "");
+
+    final SchemaRetrievalOptionsBuilder schemaRetrievalOptionsBuilder;
+    if (!databaseServerType.isUnknownDatabaseSystem()
+        && !databaseServerType
+      .getDatabaseSystemIdentifier()
+      .equalsIgnoreCase(withoutDatabasePlugin))
+    {
+      schemaRetrievalOptionsBuilder =
+        dbConnector.getSchemaRetrievalOptionsBuilder(connection);
+    }
+    else
+    {
+      schemaRetrievalOptionsBuilder = SchemaRetrievalOptionsBuilder.builder();
+    }
+
     return schemaRetrievalOptionsBuilder;
   }
 

--- a/schemacrawler-tools/src/test/java/schemacrawler/tools/databaseconnector/TestDatabaseConnectorTest.java
+++ b/schemacrawler-tools/src/test/java/schemacrawler/tools/databaseconnector/TestDatabaseConnectorTest.java
@@ -153,7 +153,7 @@ public class TestDatabaseConnectorTest
   }
 
   @Test
-  @SetSystemProperty(key = "SC_WITHOUT_DATABASE_PLUGIN", value = "true")
+  @SetSystemProperty(key = "SC_IGNORE_MISSING_DATABASE_PLUGIN", value = "true")
   public void newMajorDatabaseConnectionWithUnknownConnectorWithOverride()
     throws SchemaCrawlerException
   {

